lib/palletjack/tool.rb

```diff
@@ -0,0 +1,286 @@
+      @jack ||= PalletJack.new(options[:warehouse])
+    end
+
+    # Build a filesystem path from path components
+    #
+    # Symbols are looked up in the options dictionary.
+    # All other types are converted to strings. The
+    # resulting list is fed to File#join to produce a
+    # local filesystem compliant path.
+    #
+    # Example:
+    #   parser.on(...) {|dir| options[:output] = dir }
+    #   ...
+    #   config_path :output, 'subdir1'
+    #   config_path :output, 'subdir2'
+
+    def config_path(*path)
+      File.join(path.map {|item|
+                  case item
+                  when Symbol
+                    options.fetch(item)
+                  else
+                    item.to_s
+                  end
+                })
+    end
+
+    # :call-seq:
+    # config_dir '/path/name'
+    # config_dir :option, 'subdir', ...
+    #
+    # Creates a directory if it doesn't already exist.
+    #
+    # Uses config_path to construct the path.
+    #
+    # Example:
+    #
+    #   config_dir :output, system.name
+
+    def config_dir(*path)
+      Dir.mkdir(config_path(*path))
+    rescue Errno::EEXIST
+      nil
+    end
+
+    # :call-seq:
+    # config_file "filename" {|file| ... }
+    # config_file :option, 'fragment', 'base.ext' {|file| ... }
+    # config_file ..., mode:0600 {|file| ...}
+    #
+    # Creates a configuration file, with default mode:0644
+    # and calls the given block with the file as argument.
+    #
+    # Uses config_path to construct the path.
+    #
+    # Example:
+    #
+    #   config_file :output, system.name, 'dump.yaml' do |file|
+    #     file << system.to_yaml
+    #   end
+
+    def config_file(*path, mode:0644, &block)
+      File.open(config_path(*path),
+                File::CREAT | File::TRUNC | File::WRONLY, mode) do |file|
+        block.call(file)
+      end
+    end
+
+    # Create a new pallet dir if needed
+
+    def pallet_dir(kind, name)
+      config_dir :warehouse, kind
+      config_dir :warehouse, kind, name
+    end
+
+    # Write a new key box inside a pallet
+    #
+    # The block should return a hash representing the contents of the box.
+    # All keys will be stringified, so we can use key: short forms for
+    # declaration of the box contents.
+    #
+    # Example:
+    #   pallet_box 'domain', :domain, 'dns' do
+    #     { dns:{ ns:options[:soa_ns].split(',') } }
+    #   end
+
+    def pallet_box(kind, name, box, &block)
+      config_file :warehouse, kind, name, "#{box}.yaml" do |box_file|
+        box_file = block.call.deep_stringify_keys.to_yaml
+      end
+    end
+
+    # Create links from a pallet to parents
+    #
+    # links contain key-value pairs <link_type>:[<parent_kind>, <parent_name>]
+    #
+    # If the link target is empty (e.g. <link_type>:[]), the link is removed.
+    #
+    # Example:
+    #
+    #   pallet_links 'system', :system, os:['os', :os], netinstall:[]
+
+    def pallet_links(kind, name, links={})
```

Comment:
This is not the world's most obvious API. Then again, I'm not sure I can think of something better.

I was considering something like `{ "source_name" => { "dest_kind" => "dest_object" } }` just to get away from the mixed mess of symbols, hashes, strings and lists in the current design, but haven't actually implemented it to see what it would look like.